eth/stagedsync/stage_cumulative_index.go

```diff
@@ -0,0 +1,160 @@
+	useExternalTx := tx != nil
+
+	if !useExternalTx {
+		var err error
+		tx, err = cfg.db.BeginRw(context.Background())
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback()
+	}
+	// Log timer
+	logEvery := time.NewTicker(logInterval)
+	defer logEvery.Stop()
+
+	headNumber, err := stages.GetStageProgress(tx, stages.Headers)
+	if err != nil {
+		return fmt.Errorf("getting bodies progress: %w", err)
+	}
+	// If we are done already, we can exit the stage
+	if s.BlockNumber == headNumber {
+		return nil
+	}
+
+	cumulativeGasUsed, err := rawdb.ReadCumulativeGasUsed(tx, s.BlockNumber)
+	if err != nil {
+		return err
+	}
+
+	canonicalHashes := make(map[uint64]common.Hash)
+	// Load canonical chains
+	canonicalC, err := tx.Cursor(kv.HeaderCanonical)
+	if err != nil {
+		return err
+	}
+
+	for k, v, err := canonicalC.Seek(dbutils.EncodeBlockNumber(s.BlockNumber)); k != nil; k, v, err = canonicalC.Next() {
+		if err != nil {
+			return err
+		}
+
+		blockNum, err := dbutils.DecodeBlockNumber(k)
+		if err != nil {
+			return err
+		}
+		canonicalHashes[blockNum] = common.BytesToHash(v)
+	}
+	prevProgress := s.BlockNumber
+	headerC, err := tx.Cursor(kv.Headers)
+	if err != nil {
+		return err
+	}
+
+	currentBlockNumber := s.BlockNumber + 1
+	for k, v, err := headerC.Seek(dbutils.EncodeBlockNumber(s.BlockNumber)); k != nil; k, v, err = headerC.Next() {
+		if err != nil {
+			return err
+		}
+
+		if len(k) != 40 {
+			continue
+		}
+
+		currentBlockNumber, err = dbutils.DecodeBlockNumber(k[:8])
+		if err != nil {
+			return err
+		}
+
+		if canonicalHashes[currentBlockNumber] != common.BytesToHash(k[8:]) {
+			continue
+		}
+
+		var header types.Header
+		if err := rlp.Decode(bytes.NewReader(v), &header); err != nil {
+			log.Error("Invalid block header RLP", "number", currentBlockNumber, "err", err)
+			return nil
+		}
+		cumulativeGasUsed.Add(cumulativeGasUsed, big.NewInt(int64(header.GasUsed)))
+
+		if err := rawdb.WriteCumulativeGasUsed(tx, currentBlockNumber, cumulativeGasUsed); err != nil {
+			return err
+		}
+
+		// Sleep and check for logs
+		timer := time.NewTimer(1 * time.Nanosecond)
+		select {
+		case <-logEvery.C:
+			log.Info(fmt.Sprintf("[%s] Wrote Cumulative Index", s.LogPrefix()),
+				"gasUsed", cumulativeGasUsed.String(), "now", currentBlockNumber, "blk/sec", float64(currentBlockNumber-prevProgress)/float64(logInterval/time.Second))
+			prevProgress = currentBlockNumber
+		case <-timer.C:
+			log.Trace("RequestQueueTime (header) ticked")
+		}
+		// Cleanup timer
+		timer.Stop()
+	}
+	if err = s.Update(tx, currentBlockNumber); err != nil {
+		return err
+	}
+	if !useExternalTx {
+		if err = tx.Commit(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func UnwindCumulativeIndexStage(u *UnwindState, tx kv.RwTx, ctx context.Context) (err error) {
+	useExternalTx := tx != nil
+
+	if err = u.Done(tx); err != nil {
+		return fmt.Errorf(" reset: %w", err)
+	}
+	if !useExternalTx {
+		if err = tx.Commit(); err != nil {
```

Comment:
likely you forgot to begin tx

Reply:
It is at the beggining of the stage: ```	if !useExternalTx {
		var err error
		tx, err = cfg.db.BeginRw(context.Background())
		if err != nil {
			return err
		}
		defer tx.Rollback()
	}```

Reply:
- Unwind happens before forward
- What .begin inside Forward func will be commit/rollback inside Forward func (can't re-used it outside)
- Unwind may be run independently by ./cmd/integration